### PR TITLE
Connectathon script

### PR DESCRIPTION
**Pull request by sarahmcdougall**
# Summary
A new script was created that inserts the measure bundles from the Connectathon repository into our MongoDB database.

## New behavior
We can now insert all the measure bundles from the Connectathon repository into our database, which we can utilize in future tasks.

## Code changes
The `package.json` file was updated so that we can now use the command `npm run connectathon-upload` to run the `getConnectathonBundles.js` script. This script makes use of a gitignored connectathon repository and retrieves all the EXM bundle `JSON` files using the `getBundleFiles` function. Then, the script connects to the database, parses through each bundle file, and uploads the bundles’ resources to the appropriate collections using the `createResource` function from the mongo controller. 

# Testing guidance
`Git clone` the `connectathon` repository into the root directory of the `deqm-test-server` repository. Then, run `npm run connectathon-upload`, which will upload all the resources from the EXM bundles into the `deqm-test-server` database. Use Robo 3T to check that the appropriate collections (Patient, Encounter, ValueSet, Condition, etc.) are now populated. Note that some of the bundles make use of the same ValueSets and Libraries, so you should run into duplicate key errors in the terminal.

### CODE REVIEW

File: src/scripts/getConnectathonBundles.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+/**
+ * Uploads all the resources from the connectathon measure bundles into the
+ * database.
+ *
+ * TODO: Currently configured for fhir401 connectathon measure bundles,
+ * but may want to expand to other measure bundle providers in the future.
+ */
+async function main() {
+  const bundleFiles = getBundleFiles(connectathonPath);
```
Comment (natjoe4) [inline src/scripts/getConnectathonBundles.js:40]: I think it's a little confusing to overwrite this global variable here. Consider removing the return statement from getBundleFiles and then just calling getBundleFiles here and accessing the global bundleFiles variable which should be the same.

File: src/scripts/getConnectathonBundles.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  await mongoUtil.client.connect();
+  console.log('Connected successfully to server');
+
+  const bundlePromises = bundleFiles.map(async filePath => {
+    // read each EXM bundle file
+    const data = fs.readFileSync(filePath, 'utf8');
+    if (data) {
+      const bundle = await JSON.parse(data);
```
Comment (natjoe4) [inline src/scripts/getConnectathonBundles.js:49]:
```suggestion
      const bundle = JSON.parse(data);
```
JSON.parse is synchronous so this await is unnecessary

**Review by natjoe4 — CHANGES_REQUESTED** [review]
This looks good to me and the functionality was as advertised. I left a couple of very small comments. I'm still pretty shaky on async logic inside map functions, so I'd recommend having one more person look that over if you haven't already. Good work!

File: src/scripts/getConnectathonBundles.js (hunk 1 above)
Comment (sarahmcdougall) [reply src/scripts/getConnectathonBundles.js:40]: Good catch, I think I was renaming stuff and didn't realize I did this

**Review by natjoe4 — APPROVED** [review]

File: src/scripts/getConnectathonBundles.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * Uploads all the resources from the connectathon measure bundles into the
+ * database.
+ *
+ * TODO: Currently configured for fhir401 connectathon measure bundles,
+ * but may want to expand to other measure bundle providers in the future.
+ */
+async function main() {
+  getBundleFiles(connectathonPath);
```
Comment (hossenlopp) [inline src/scripts/getConnectathonBundles.js:39]: Suggest adding a check to make sure the c-thon folder exists first. Giving a helpful error message to do the clone before running this script would be a nice addition.

File: src/scripts/getConnectathonBundles.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    // read each EXM bundle file
+    const data = fs.readFileSync(filePath, 'utf8');
+    if (data) {
+      const bundle = JSON.parse(data);
+      // retrieve each resource and insert into database
+      const uploads = bundle.entry.map(async res => {
+        try {
+          await createResource(res.resource, res.resource.resourceType);
+        } catch (e) {
+          console.log(e.message);
```
Comment (hossenlopp) [inline src/scripts/getConnectathonBundles.js:54]: It's very common that the shared Libraries or ValueSets will be in multiple measure bundles and will fail to insert (assuming indexes are configured) after the first copy is inserted. This is a normal thing that shouldn't spam the output. Either catch this `duplicate key error` and not bother printing it out unless it is a `Measure` resource or check for existence before attempting the insert.

**Review by natjoe4 — APPROVED** [review]

**Review by hossenlopp — CHANGES_REQUESTED** [review]
Async stuff looks all correct and it works. Just have some concerns about the error states.

File: src/scripts/getConnectathonBundles.js (hunk 3 above)
Comment (sarahmcdougall) [reply src/scripts/getConnectathonBundles.js:39]: Good point! That should definitely be added

**Review by natjoe4 — APPROVED** [review]

File: src/scripts/getConnectathonBundles.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    const data = fs.readFileSync(filePath, 'utf8');
+    if (data) {
+      const bundle = JSON.parse(data);
+      // retrieve each resource and insert into database
+      const uploads = bundle.entry.map(async res => {
+        try {
+          await createResource(res.resource, res.resource.resourceType);
+        } catch (e) {
+          // ignore duplicate key errors for Libraries, ValueSets
+          if (res.resource.resourceType === 'Measure') {
```
Comment (hossenlopp) [inline src/scripts/getConnectathonBundles.js:61]: This hides ALL errors unless they are on Measure. Suggest checking on error code. `11000` is the code for duplicate key.
```suggestion
          if (e.code !== 11000 || res.resource.resourceType === 'Measure') {
```

**Review by hossenlopp — CHANGES_REQUESTED** [review]
Shouldn't filter out ALL error messages if they aren't on `Measure`.

File: src/scripts/getConnectathonBundles.js (hunk 5 above)
Comment (sarahmcdougall) [reply src/scripts/getConnectathonBundles.js:61]: Thank you!!

**Review by hossenlopp — APPROVED** [review]
Looks good!

**Review by natjoe4 — APPROVED** [review]

**Pull request merged**